workers: add docs for getBindingsProxy utility (#12441)
cloudflare/cloudflare-docs: content/workers/wrangler/api.md

# Wrangler API

Wrangler offers APIs to programmatically interact with your Cloudflare Workers.

- [`unstable_dev`](#unstable_dev) - Start a server for running either end-to-end (e2e) or integration tests against your Worker.

> - [`getBindingsProxy`](#getBindingsProxy) - get bindings via Wrangler and use them in your code.

Reviewer: ```suggestion
- [`getBindingsProxy`](#getBindingsProxy) - Get bindings via Wrangler and use them in your code.
```

{{</tab>}}
{{</tabs>}}

## `getBindingsProxy`

This function is used to get a proxy for **local** `workerd` bindings that can be then used in code running via Node.js processes for Workers and Pages projects.

### Usage

The function uses bindings found in `wrangler.toml`. For example, if you have this in `wrangler.toml`:

```js
[vars]

MY_VARIABLE = "test"
```

You can access the bindings by importing `getBindingsProxy` like this:

```js
import { getBindingsProxy } from "wrangler";

const { bindings } = await getBindingsProxy();
```

To access the value of the `MY_VARIABLE` binding add the following to your code:

```js
console.log(`MY_VARIABLE = ${bindings['MY_VARIABLE']}`);
```

You should see the returned value: `test`.

### Supported bindings

> All supported bindings found in your `wrangler.toml` are available to you via `bindings`. Supported bindings currently include: kv, r2, d1, service, queue-producer, and var.

Reviewer: ```suggestion
All supported bindings found in your `wrangler.toml` are available to you via `bindings`.

The bindings supported by `getBindingsProxy` are:
 - [Environmental variables](/workers/wrangler/configuration/#environmental-variables)
 - [Service bindings](/workers/configuration/bindings/#service-bindings)
 - [KV namespace bindings](/workers/configuration/bindings/#kv-namespace-bindings)
 - [Durable Object bindings](/workers/configuration/bindings/#durable-object-bindings)
 - [R2 bucket bindings](/workers/configuration/bindings/#r2-bucket-bindings)
 - [D1 database bindings](/workers/configuration/bindings/#d1-database-bindings)
  - [R2 bucket bindings](https://developers.cloudflare.com/workers/configuration/bindings/#r2-bucket-bindings)
  - [D1 database bindings](https://developers.cloudflare.com/workers/configuration/bindings/#d1-database-bindings)
```

{{</tab>}}
{{</tabs>}}

## `getBindingsProxy`

This function is used to get a proxy for **local** `workerd` bindings that can be then used in code running via Node.js processes for Workers and Pages projects.

### Usage

> The function uses bindings found in `wrangler.toml`. For example, if you have this in `wrangler.toml`:

Reviewer 2: Making this more clear as to what type of binding this is.

{{</tab>}}
{{</tabs>}}

## `getBindingsProxy`

This function is used to get a proxy for **local** `workerd` bindings that can be then used in code running via Node.js processes for Workers and Pages projects.

### Usage

The function uses bindings found in `wrangler.toml`. For example, if you have this in `wrangler.toml`:

```js
[vars]

MY_VARIABLE = "test"
```

You can access the bindings by importing `getBindingsProxy` like this:

```js
import { getBindingsProxy } from "wrangler";

const { bindings } = await getBindingsProxy();
```

To access the value of the `MY_VARIABLE` binding add the following to your code:

```js
console.log(`MY_VARIABLE = ${bindings['MY_VARIABLE']}`);
```

> You should see the returned value: `test`.

Reviewer: ```suggestion
In your terminal you should see: `MY_VARIABLE = test`
```

> 

Reviewer: ```suggestion
```

{{</tab>}}
{{</tabs>}}

## `getBindingsProxy`

> This function is used to get a proxy for **local** `workerd` bindings that can be then used in code running via Nodejs processes for Workers and Pages projects.

Reviewer: ```suggestion
This function is used to get a proxy for **local** `workerd` bindings that can be then used in code running via Node.js processes for Workers and Pages projects.
```

### Usage

The function uses bindings found in `wrangler.toml`. For example, if you have this in `wrangler.toml`:

```js
[vars]

MY_VARIABLE = "test"
```

You can access the bindings by importing `getBindingsProxy` like this:

```js
import { getBindingsProxy } from "wrangler";

const { bindings } = await getBindingsProxy();
```

To access the value of the `MY_VARIABLE` binding add the following to your code:

```js
console.log(`MY_VARIABLE = ${bindings['MY_VARIABLE']}`);
```

You should see the returned value: `test`.

> {{<Aside type="note">}}
> 
> For a demo of this API, see [getBindingsProxy-poc-demo](https://github.com/dario-piotrowicz/getBindingsProxy-poc-demo).
> 
> {{</Aside>}}

Reviewer: ```suggestion
```

My demo there is really low quality (I wasn't planning to share it in docs) and it's also under my user name

I think we should avoid this, hopefully the above info is enough, if we would also really want to add a demo I can look into making a better one, potentially under the Cloudflare org to share here 🙂

Author: That's fine. I always want to give more examples when we can 👍

### Supported bindings

All supported bindings found in your `wrangler.toml` are available to you via `bindings`. Supported bindings currently include: kv, r2, d1, service, queue-producer, and var.

> {{<Aside type="note">}}
> 
> While this function works for both Workers and Pages projects, Durable Objects integration has not been implemented for Pages.
> 
> {{</Aside>}}

Reviewer: ```suggestion
```

`getBindingsProxy` does work with Durable Objects 🙂

The only thing is that, since in Pages applications you can't define Durable Objects, the only way to integrate with DO bindings is by having workers exposing DurableObjects run with `wrangler dev` and reference those.

For example see the DO bindings usage [here](
https://github.com/cloudflare/workers-sdk/blob/9f96f28b88252dc62f1901f6533a69218f96c2dd/fixtures/get-bindings-proxy/tests/index.test.ts#L116-L124) , where this is the respective [toml DO section](https://github.com/cloudflare/workers-sdk/blob/9f96f28b88252dc62f1901f6533a69218f96c2dd/fixtures/get-bindings-proxy/wrangler.toml#L22-L26) and these two ([a](https://github.com/cloudflare/workers-sdk/tree/9f96f28b88252dc62f1901f6533a69218f96c2dd/fixtures/get-bindings-proxy/workers/do-worker-a) and [b](https://github.com/cloudflare/workers-sdk/tree/9f96f28b88252dc62f1901f6533a69218f96c2dd/fixtures/get-bindings-proxy/workers/do-worker-b)) are the workers defining the DOs 🙂 

`wrangler pages dev` supports DOs in the same manner: https://developers.cloudflare.com/pages/functions/bindings/#interact-with-your-durable-object-namespaces-locally

Author: Ok. This is good to be removed. Thanks for the examples!
